src/utils/__tests__/extractPaginationParameters.unit.test.ts

```diff
@@ -0,0 +1,49 @@
+import {
+  extractPaginationParameters,
+} from '..';
+import {
+  InputValidationError,
+} from '../../errors';
+
+describe('extractPaginationParameters', () => {
+  it('should return default values according to documented defaults', () => {
+    const paginationParameters = extractPaginationParameters({
+      query: {},
+    });
+    expect(paginationParameters).toEqual({
+      page: 1,
+      count: 100,
```

Comment:
Possibly... we haven't been treating openapi.json as a form of documentation less than a driver of functionality (part of this is because typescript is not great at importing types from json files).

Given that, this would be the first (and only) formal coupling.  It might be surprising behavior for future developers to have *just this* aspect of the openapi spec be a direct driver of code, when nothing else does (I know there's a risk of stale documentation, but again that is a constant risk for all other aspects of the openAPI spec right now).

So the question is: is the DRY / staleness protection worth the "this is different than all the other API functionality's relationship with the spec"?

Reply:
I'll take a look at the oath config process and see if I can leverage that approach.

The bigger question of how we can connect the two in a more formal way is something that I think will warrant a separate issue.  Right now we're defining *every schema* twice -- once in the spec, once in our type.   I spent a few cycles early on trying to find a better way but they all felt pretty brittle at the time.  Worth revisiting.

Reply:
Here's the relevant code: https://github.com/PhilanthropyDataCommons/service/blob/bb25a75dcb218aca886435f5449f39ff6a76d6ee/src/routers/documentationRouter.ts#L36-L45

I thought about suggesting some kind of deep merge of a nested object, but there isn't great first-class support for that in JavaScript and it didn't seem worthwhile to bring in a new dependency specifically for that. If we'll be doing more in that direction, though, maybe it is worthwhile!

> Right now we're defining _every schema_ twice -- once in the spec, once in our type.

I think it is worth differentiating values from types here; I saw you doing some of that work, and I've done similar work before, and yeah unifying types is very very difficult. Default values, though - and whatever other hardcoded constants we might have - seem much more achievable!

Reply:
Ahh.... digging into this now.

I see why this approach makes lots of sense for a value that *will* be necessarily overriden by a local copy, but I'm not sure that's what's going on here.  I don't think we want these values to be configurable by the server, right?  e.g. should it be that the PDC API might have different defaults API values in different implementations?  It feels to me that default values are actually driven by specification, not by the server.

So while I could have this driven by a constant / env var, that constant should never change between instances.

Meanwhile, we want our spec to be a useful artefact without having to run the server, so I'm going to want to provide the value in the `openapi.json` as well (We want someone to be able to just download the swagger spec from our repo and throw it into their favorite tool / use it as an artefact in their client / etc.)

This means we'd still have redundancy.

All this said I'll see if I can just load it from the swagger spec as you described; to your point this is not the same as a type so I think we should be able to more safely define and access it.  Will circle back with an update in a moment!

Reply:
^^
Using these values for the tests are going to get a little weird.

The tests will be *entirely* dynamic, and start to become a bit obscured in logic.  I feel we might be taking steps back here in long term legibility in the name of avoiding redundancy.  When these values change it's a change to the API spec -- just like if we were to rename a parameter name, or a path name, etc.

This might be a strawman, but we don't have parameter names driven by the spec -- it's a dev / review task to make sure the implementation is updated to reflect the spec!  Maybe this is the same concept?

Reply:
UPDATE: it's super trivial to pull out the value from the spec (the thing that was not trivial was pulling schema definitions / types.)

Reply:
One note: I'm a bit worried that using the same `apiSpecification.parameters.pageParam.default` value for both the implementation *and* the test is brittle.  This means the test would fail if the api spec was somehow incorrect.

Ultimately by loading the spec in our implementation we are making the spec part of the implementation -- that means if we load the spec into the test we're writing a test that is defined in terms of the implementation, making it possible to do something like break the spec in a way that breaks the implementation and have the tests still pass.

Would it be OK for me to have the test hard coded? There's no real risk of the test becoming out of date since it will be loud if the spec ever changes without a corresponding test update.

Reply:
> UPDATE: it's super trivial to pull out the value from the spec

Yay! I'm glad!

> Ultimately by loading the spec in our implementation we are making the spec part of the implementation -- that means if we load the spec into the test we're writing a test that is defined in terms of the implementation, making it possible to do something like break the spec in a way that breaks the implementation and have the tests still pass.

It sounds like you're describing a need for tests for the spec! I think there's value in the tests validating that the default value comes from the spec, and perhaps there's also value in having tests that validate that the values in the spec are reasonable - like that they are non-negative numbers, or whatever.

Reply:
Hmmm fair -- and I suppose the main case I'm worried about (param renaming) would cause a test failure because `apiSpecification.parameters.pageParam.default` would now be trying to get the `default` attribute of a null value, so the test would simply error.

FWIW: I do agree with you Jesse that it'd be better to just make this spec a build artifact in the long run.

Reply:
Tests updated (and spec tests added!)

Reply:
> On the issue of whether to generate docs from code or code from docs, I prefer the former. [...] the creation of an `openapi.json` would be a build step [...] with the resulting documentation included in the build artifact.

This sounds plausible to me; can we talk about the conditions that would lead to that? Static assets are simpler (both computationally and conceptually) than built assets, so I think it's worthwhile to avoid that if we can. Does the auth stuff merit this step, or are we still okay with the local overrides at run time? If not auth stuff, what kind of dynamism are you imagining would make us want to construct the OpenAPI docs at build time?

Reply:
It's not dynamism as much as specifying everything twice: once in the code and once in the docs. It's really a separate issue from this, though, and the ship may have already sailed.